### PR TITLE
fix(typescript): incremental typescript cache (#859)

### LINKED ISSUE #698: Incremental build not working
- Rollup Plugin Name: @rollup/plugin-typescript
- Rollup Plugin Version: 8.0.0
- Rollup Version: 2.34.2
- Operating System (or Browser): Win10
- Node Version: v10.16.0
- Link to reproduction _(⚠️ read below)_: https://github.com/cdalexndr/rollup-plugin-issue-698

### Expected Behavior
Compilation successfull.

### Actual Behavior
`Error: @rollup/plugin-typescript: Rollup 'dir' option must be used when Typescript compiler options 'tsBuildInfoFile' or 'incremental' are specified.`

### Additional Information
I have a working build configuration that outputs bundles.

When I add `incremental` to my `tsconfig.json`, the build fails with:
`Error: @rollup/plugin-typescript: Rollup 'dir' option must be used when Typescript compiler options 'tsBuildInfoFile' or 'incremental' are specified.`

If I add `dir: ".rollup.dir"` to my `rollup.config.js`, then build fails with another error:
`Error: You must set either "output.file" for a single-file build or "output.dir" when generating multiple chunks.`

Isn't incremental build supported? I expected this feature to be plug and play.

### PR DESCRIPTION
**Pull request by shellscape**
## Rollup Plugin Name: `typescript`

This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

If yes, then include "BREAKING CHANGES:" in the first commit message body, followed by a description of what is breaking. 

List any relevant issue numbers: Fixes #698

### Description

This PR continues the work started by #859.

### CODE REVIEW

File: packages/typescript/src/outputFile.ts [hunk 1]
```diff
@@ -72,3 +76,36 @@ export default function findTypescriptOutput(
...
+
+export async function emitFile(
+  outputOptions: OutputOptions,
+  outputToFilesystem: boolean | undefined,
+  context: PluginContext,
+  filePath: string,
+  fileSource: string
+) {
+  // TODO: this is confusing, change
+  const fromRollupOutputDirToFile = outputOptions.dir
```
Comment (NotWoods) [inline packages/typescript/src/outputFile.ts:92]: This line can probably be combined with the if statement, it only is used in one branch.

```ts
if (!outputOptions.dir) {
  if (outputToFilesystem == null)...
} else {
  const fromRollupOutputDirToFile = path.relative(outputOptions.dir, filePath)
  context.emitFile({ ...
}
```

**Review by NotWoods — APPROVED** [review]
good to go, small optional change

File: packages/typescript/src/outputFile.ts (hunk 1 above)
Comment (shellscape) [reply packages/typescript/src/outputFile.ts:92]: That change results in a failing test:

```
  Error {
    code: 'PLUGIN_ERROR',
    hook: 'generateBundle',
    plugin: 'typescript',
    pluginCode: 'VALIDATION_ERROR',
    message: 'The "fileName" or "name" properties of emitted files must be strings that are neither absolute nor relative paths and do not contain invalid characters, received "../.tsbuildinfo".',
  }
 ```
 
 I'll see if I can refactor that in another way.

**Comment by shellscape** [comment]
@NotWoods I refactored that block a bit. But overall the logic has to remain the same. There's a bit of an edge case in situations where `path.relative` returns a path prefixed with `..` and the else case for the lack of `outputOptions.dir` has to mimic that. Super funky, but necessary. Please take a second look before I merge this.

**Review by NotWoods — APPROVED** [review]
LGTM

**Pull request merged**